### PR TITLE
Add telemetry

### PR DESCRIPTION
**Pull request by ianagne**
Add the following telemetry events surrounding HTTP requests:
[:algolia, :request, :start]
[:algolia, :request, :stop]
[:algolia, :request, :exception]

In addition to the standard metadata (e.g. duration), we will emit the
following attributes:
* `status` - Returns `ok` on success, `error` on failure
* `attempt` - The current number of the attempt. The HTTP request will
  be retried on failure.
* `method` - The HTTP method (e.g. :get, :post) of the request
* `status_code` - The HTTP status code of the response
* `url` - The request url

### CODE REVIEW

File: test/algolia_test.exs [hunk 1]
```diff
@@ -4,44 +4,73 @@ defmodule AlgoliaTest do
...
-      "test_1"
-      |> add_objects([%{text: "add multiple test"}, %{text: "add multiple test"}, %{text: "add multiple test"}])
-      |> wait
+             "test_1"
+             |> add_objects([
+               %{text: "add multiple test"},
+               %{text: "add multiple test"},
+               %{text: "add multiple test"}
+             ])
+             |> wait
```
Comment (elainenaomi) [inline test/algolia_test.exs]: nit
```suggestion
             |> wait()
```

File: test/algolia_test.exs [hunk 2]
```diff
@@ -4,44 +4,73 @@ defmodule AlgoliaTest do
...
     |> Enum.each(&wait/1)
   end
 
   test "add object" do
+    TelemetryHelper.attach_telemetry([:algolia, :request])
+
     {:ok, %{"objectID" => object_id}} =
       "test_1"
       |> add_object(%{text: "hello"})
       |> wait
```
Comment (elainenaomi) [inline test/algolia_test.exs]: nit
```suggestion
      |> wait()
```

File: test/algolia_test.exs [hunk 3]
```diff
@@ -50,8 +79,9 @@ defmodule AlgoliaTest do
   end
 
   test "wait task" do
-    :rand.seed(:exs1024, :erlang.timestamp)
-    object_id = :rand.uniform(1000000) |> to_string
+    :rand.seed(:exs1024, :erlang.timestamp())
+    object_id = :rand.uniform(1_000_000) |> to_string
```
Comment (elainenaomi) [inline test/algolia_test.exs]: nit
```suggestion
    object_id = :rand.uniform(1_000_000) |> to_string()
```

File: test/algolia_test.exs [hunk 4]
```diff
@@ -61,8 +91,8 @@ defmodule AlgoliaTest do
   end
 
   test "save one object, and then read it, using wait_task pipeing" do
-    :rand.seed(:exs1024, :erlang.timestamp)
-    id = :rand.uniform(1000000) |> to_string
+    :rand.seed(:exs1024, :erlang.timestamp())
+    id = :rand.uniform(1_000_000) |> to_string
```
Comment (elainenaomi) [inline test/algolia_test.exs]:
```suggestion
    id = :rand.uniform(1_000_000) |> to_string()
```

File: test/algolia_test.exs [hunk 5]
```diff
@@ -61,8 +91,8 @@ defmodule AlgoliaTest do
   end
 
   test "save one object, and then read it, using wait_task pipeing" do
```
Comment (elainenaomi) [inline test/algolia_test.exs]:
```suggestion
  test "save one object, and then read it, using wait_task piping" do
```

File: test/algolia_test.exs [hunk 6]
```diff
@@ -87,26 +117,27 @@ defmodule AlgoliaTest do
   end
 
   test "search single index" do
-    :rand.seed(:exs1024, :erlang.timestamp)
-    count = :rand.uniform 10
-    docs = Enum.map(1..count, &(%{id: &1, test: "search_single_index"}))
+    :rand.seed(:exs1024, :erlang.timestamp())
+    count = :rand.uniform(10)
+    docs = Enum.map(1..count, &%{id: &1, test: "search_single_index"})
```
Comment (elainenaomi) [inline test/algolia_test.exs:122]: I don't remember if we could add an extra space here (just to improve readability)
```suggestion
    docs = Enum.map(1..count, & %{id: &1, test: "search_single_index"})
```

File: test/algolia_test.exs [hunk 7]
```diff
@@ -87,26 +117,27 @@ defmodule AlgoliaTest do
...
 
-    {:ok, _} =  save_objects("test_3", docs, id_attribute: :id) |> wait
+    {:ok, _} = save_objects("test_3", docs, id_attribute: :id) |> wait
```
Comment (elainenaomi) [inline test/algolia_test.exs]:
```suggestion
    {:ok, _} = save_objects("test_3", docs, id_attribute: :id) |> wait()
```

File: test/algolia_test.exs [hunk 8]
```diff
@@ -87,26 +117,27 @@ defmodule AlgoliaTest do
...
     assert length(hits1) === count
   end
 
   test "search with list opts" do
-    :rand.seed(:exs1024, :erlang.timestamp)
-    count = :rand.uniform 10
-    docs = Enum.map(1..count, &(%{id: &1, test: "search with list opts"}))
+    :rand.seed(:exs1024, :erlang.timestamp())
+    count = :rand.uniform(10)
+    docs = Enum.map(1..count, &%{id: &1, test: "search with list opts"})
```
Comment (elainenaomi) [inline test/algolia_test.exs:133]: same comment as above
```suggestion
    docs = Enum.map(1..count, & %{id: &1, test: "search with list opts"})
```

File: test/algolia_test.exs [hunk 9]
```diff
@@ -115,7 +146,7 @@ defmodule AlgoliaTest do
   end
 
   test "search > 1 pages" do
-    docs = Enum.map(1..40, &(%{id: &1, test: "search_more_than_one_pages"}))
+    docs = Enum.map(1..40, &%{id: &1, test: "search_more_than_one_pages"})
```
Comment (elainenaomi) [inline test/algolia_test.exs:149]: same comment as above
```suggestion
    docs = Enum.map(1..40, & %{id: &1, test: "search_more_than_one_pages"})
```

File: test/algolia_test.exs [hunk 10]
```diff
@@ -115,7 +146,7 @@ defmodule AlgoliaTest do
...
 
     {:ok, _} = save_objects("test_3", docs, id_attribute: :id) |> wait
```
Comment (elainenaomi) [inline test/algolia_test.exs]:
```suggestion
    {:ok, _} = save_objects("test_3", docs, id_attribute: :id) |> wait()
```

File: test/algolia_test.exs [hunk 11]
```diff
@@ -87,26 +117,27 @@ defmodule AlgoliaTest do
...
 
   test "search with list opts" do
-    :rand.seed(:exs1024, :erlang.timestamp)
-    count = :rand.uniform 10
-    docs = Enum.map(1..count, &(%{id: &1, test: "search with list opts"}))
+    :rand.seed(:exs1024, :erlang.timestamp())
+    count = :rand.uniform(10)
+    docs = Enum.map(1..count, &%{id: &1, test: "search with list opts"})
 
     {:ok, _} = save_objects("test_3", docs, id_attribute: :id) |> wait
```
Comment (elainenaomi) [inline test/algolia_test.exs]:
```suggestion
    {:ok, _} = save_objects("test_3", docs, id_attribute: :id) |> wait()
```

File: test/algolia_test.exs [hunk 12]
```diff
@@ -165,15 +196,23 @@ defmodule AlgoliaTest do
...
+             }
+           ] == hits
   end
 
   defp generate_fixtures_for_index(index) do
-    :rand.seed(:exs1024, :erlang.timestamp)
+    :rand.seed(:exs1024, :erlang.timestamp())
     count = :rand.uniform(3)
-    objects = Enum.map(1..count, &(%{objectID: &1, test: "search_multiple_indexes"}))
+    objects = Enum.map(1..count, &%{objectID: &1, test: "search_multiple_indexes"})
```
Comment (elainenaomi) [inline test/algolia_test.exs:215]:
```suggestion
    objects = Enum.map(1..count, & %{objectID: &1, test: "search_multiple_indexes"})
```

File: test/algolia_test.exs [hunk 13]
```diff
@@ -197,41 +236,63 @@ defmodule AlgoliaTest do
     id = "partially_update_object_upsert_true"
 
     assert {:ok, _} =
-      partial_update_object("test_2", %{}, id)
-      |> wait
+             partial_update_object("test_2", %{}, id)
+             |> wait
```
Comment (elainenaomi) [inline test/algolia_test.exs]:
```suggestion
             |> wait()
```

File: test/algolia_test.exs [hunk 14]
```diff
@@ -197,41 +236,63 @@ defmodule AlgoliaTest do
...
   test "partial update object, upsert is false" do
+    TelemetryHelper.attach_telemetry([:algolia, :request])
+
     id = "partial_update_upsert_false"
 
     assert {:ok, _} =
-      partial_update_object("test_3", %{update: "updated"}, id, upsert?: false)
-      |> wait
+             partial_update_object("test_3", %{update: "updated"}, id, upsert?: false)
+             |> wait
```
Comment (elainenaomi) [inline test/algolia_test.exs]:
```suggestion
             |> wait()
```

File: test/algolia_test.exs [hunk 15]
```diff
@@ -197,41 +236,63 @@ defmodule AlgoliaTest do
...
 
     {:ok, object} = get_object("test_2", id)
     assert object["objectID"] == id
   end
 
   test "partial update object, upsert is false" do
+    TelemetryHelper.attach_telemetry([:algolia, :request])
+
     id = "partial_update_upsert_false"
```
Comment (elainenaomi) [inline test/algolia_test.exs]: maybe we could use an ID like this to improve the test readability
```suggestion
    id = "non_existent_id_to_update"
```

File: test/algolia_test.exs [hunk 16]
```diff
@@ -197,41 +236,63 @@ defmodule AlgoliaTest do
...
   end
 
   test "partially update multiple objects, upsert is default" do
     objects = [%{id: "partial_update_multiple_1"}, %{id: "partial_update_multiple_2"}]
 
     assert {:ok, _} =
-      partial_update_objects("test_3", objects, id_attribute: :id)
-      |> wait
+             partial_update_objects("test_3", objects, id_attribute: :id)
+             |> wait
```
Comment (elainenaomi) [inline test/algolia_test.exs]:
```suggestion
             |> wait()
```

File: test/algolia_test.exs [hunk 17]
```diff
@@ -197,41 +236,63 @@ defmodule AlgoliaTest do
...
+    objects = [
+      %{id: "partial_update_multiple_1_no_upsert"},
+      %{id: "partial_update_multiple_2_no_upsert"}
+    ]
 
     assert {:ok, _} =
-      partial_update_objects("test_3", objects, id_attribute: :id, upsert?: false)
-      |> wait
+             partial_update_objects("test_3", objects, id_attribute: :id, upsert?: false)
+             |> wait
```
Comment (elainenaomi) [inline test/algolia_test.exs]:
```suggestion
             |> wait()
```

File: test/algolia_test.exs [hunk 18]
```diff
@@ -253,6 +314,7 @@ defmodule AlgoliaTest do
 
   test "delete multiple objects" do
     objects = [%{id: "delete_multipel_objects_1"}, %{id: "delete_multipel_objects_2"}]
+
     {:ok, %{"objectIDs" => object_ids}} =
       save_objects("test_1", objects, id_attribute: :id)
       |> wait
```
Comment (elainenaomi) [inline test/algolia_test.exs]:
```suggestion
      |> wait()
```

File: test/algolia_test.exs [hunk 19]
```diff
@@ -305,10 +367,12 @@ defmodule AlgoliaTest do
     attributesToIndex = ~w(foo bar baz)
 
     assert {:ok, _} =
-      @settings_test_index
-      |> set_settings(%{attributesToIndex: attributesToIndex})
-      |> wait
-    assert {:ok, %{"attributesToIndex" => ^attributesToIndex}} = get_settings(@settings_test_index)
+             @settings_test_index
+             |> set_settings(%{attributesToIndex: attributesToIndex})
+             |> wait
```
Comment (elainenaomi) [inline test/algolia_test.exs]:
```suggestion
             |> wait()
```

File: lib/algolia.ex [hunk 20]
```diff
@@ -186,6 +178,53 @@ defmodule Algolia do
...
+  defp do_request(method, url, headers, body, curr_retry) do
+    telemetry_metadata = %{attempt: curr_retry + 1, http: %{method: method, url: url}}
+
+    :telemetry.span([:algolia, :request], telemetry_metadata, fn ->
+      http_result =
+        :hackney.request(method, url, headers, body, [
+          :with_body,
+          path_encode_fun: &URI.encode/1,
+          connect_timeout: 3_000 * (curr_retry + 1),
+          recv_timeout: 30_000 * (curr_retry + 1),
```
Comment (elainenaomi) [inline lib/algolia.ex:190]: TIL: we have 3 retries before returning "Unable to connect to Algolia"

File: test/telemetry_helper.exs [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+  end
+
+  # Why isn't send_event/4 an anonymous function or private function? 
+  #
+  # Note from telemetry documentation: due to how anonymous functions are
+  # implemented in the Erlang VM, it is best to use function captures (i.e. fun
+  # mod:fun/4 in Erlang or &Mod.fun/4 in Elixir) as event handlers to achieve
+  # maximum performance. In other words, avoid using literal anonymous functions
+  # (fun(...) -> ... end or fn ... -> ... end) or local function captures (fun
+  # handle_event/4 or &handle_event/4 ) as event handlers.
```
Comment (elainenaomi) [inline test/telemetry_helper.exs:30]: TIL

**Comment by msramos** [comment]
We should update the `README` as well

File: lib/algolia.ex [hunk 22]
```diff
@@ -186,6 +178,53 @@ defmodule Algolia do
...
+          ssl_options: [{:versions, [:"tlsv1.2"]}],
+          pool: pool_name()
+        ])
+
+      {http_result, update_telemetry_metadata(http_result, telemetry_metadata)}
```
Comment (elainenaomi) [inline lib/algolia.ex:195]: not sure if I understood right, but it seems the response would be a tuple like:
```elixir
{{:ok, _code, _headers, _response}, %{status: 200, http: ...} = telemetry_metadata_as_map}
```

or will it be something else? 🤔 
if it's the tuple I mentioned, it seems we need to update the pattern matching in the case - line 169

**Review by elainenaomi — COMMENTED** [review]
Awesome work! 💯 💯 💯 

I added a question about the "do_request" call and I was thinking of updating the README.file with an instruction about how to run the tests locally

Wdyt?

File: lib/algolia.ex (hunk 22 above)
Comment (ianagne) [reply lib/algolia.ex:195]: Ah, this is a peculiarity of telemetry. When returning something for `:telemetry.span()`, the shape of the return is `{result, telemetry_metadata}`

The successful return out the do_request/5 function will still be something along the lines of `{:ok, _code, _headers, _response}`.

Comment (elainenaomi) [reply lib/algolia.ex:195]: aah got it :TIL:
thanks for the explanation, @ianagne ❤️ 
maybe a doc for this function would be interesting 😊

**Review by elainenaomi — APPROVED** [review]
🎉 🎉

File: test/algolia_test.exs (hunk 6 above)
Comment (ianagne) [reply test/algolia_test.exs:122]: `mix format` doesn't seem to accept this change

Comment (elainenaomi) [reply test/algolia_test.exs:122]: 💔

**Comment by ianagne** [comment]
@msramos The README.md has been updated [with a Telemetry section](https://github.com/TheRealReal/algolia-elixir/blob/04330ca121aef9da0aebf4e80672bb7c596425a6/README.md#telemetry)

**Review by msramos — APPROVED** [review]

**Pull request merged**